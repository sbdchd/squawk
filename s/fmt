#!/bin/sh
set -eu

cargo fmt
