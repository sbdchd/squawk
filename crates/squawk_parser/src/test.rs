// based on https://github.com/rust-lang/rust-analyzer/blob/d8887c0758bbd2d5f752d5bd405d4491e90e7ed6/crates/parser/src/tests.rs
use camino::Utf8Path;
use dir_test::{dir_test, Fixture};
use insta::{assert_snapshot, with_settings};
use std::fmt::Write;

use crate::{parse, LexedStr};

fn parse_text(text: &str) -> (String, bool) {
    let lexed = LexedStr::new(text);
    let input = lexed.to_input();
    let output = parse(&input);

    let mut buf = String::new();
    let mut errors = Vec::new();
    let mut indent = String::new();
    let mut depth = 0;
    let mut len = 0;
    lexed.intersperse_trivia(&output, &mut |step| match step {
        crate::StrStep::Token { kind, text } => {
            assert!(depth > 0);
            len += text.len();
            writeln!(buf, "{indent}{kind:?} {text:?}").unwrap();
        }
        crate::StrStep::Enter { kind } => {
            assert!(depth > 0 || len == 0);
            depth += 1;
            writeln!(buf, "{indent}{kind:?}").unwrap();
            indent.push_str("  ");
        }
        crate::StrStep::Exit => {
            assert!(depth > 0);
            depth -= 1;
            indent.pop();
            indent.pop();
        }
        crate::StrStep::Error { msg, pos } => {
            assert!(depth > 0);
            let err = "ERROR";
            errors.push(format!("{err}@{pos}: {msg}\n"));
        }
    });
    assert_eq!(
        len,
        text.len(),
        "didn't parse all text.\nParsed:\n{}\n\nAll:\n{}\n",
        &text[..len],
        text
    );

    for (token, msg) in lexed.errors() {
        let pos = lexed.text_start(token);
        let err = "ERROR";
        errors.push(format!("{err}@{pos}: {msg}\n"));
    }

    let has_errors = !errors.is_empty();
    if has_errors {
        buf.push_str("---\n");
        for e in errors {
            buf.push_str(&e);
        }
    }
    (buf, has_errors)
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_data",
    glob: "**/*.sql",
)]
fn sqltest(fixture: Fixture<&str>) {
    let content = fixture.content();
    let absolute_fixture_path = Utf8Path::new(fixture.path());
    let input_file = absolute_fixture_path;
    let test_name = absolute_fixture_path
        .file_name()
        .and_then(|x| x.strip_suffix(".sql"))
        .unwrap();

    let parent_dir = input_file.parent().and_then(|x| x.file_name()).unwrap();

    let (parsed, has_errors) = parse_text(content);

    with_settings!({
      omit_expression => true,
      input_file => input_file,
    }, {
      assert_snapshot!(format!("{}_{}", test_name, parent_dir), parsed);
    });

    if parent_dir == "ok" {
        // We check that all of our tests in `ok` also pass the Postgres parser,
        // if they don't, they should be moved to the `err` directory.
        assert!(
            !has_errors,
            "tests defined in the `ok` can't have parser errors."
        );
        // skipping pg17 specific stuff since our parser isn't using the latest parser
        if !test_name.ends_with("pg17") {
            let pg_result = pg_query::parse(content);
            if let Err(e) = &pg_result {
                assert!(
                    &pg_result.is_ok(),
                    "tests defined in the `ok` can't have Postgres parser errors. Found {}",
                    e
                );
            }
        }
    } else {
        assert_eq!(parent_dir, "err");
        assert!(
            has_errors,
            "tests defined in the `err` directory must have parser errors."
        );
    }
}
