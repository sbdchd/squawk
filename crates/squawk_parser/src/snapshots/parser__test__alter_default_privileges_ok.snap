---
source: crates/squawk_parser/src/test.rs
input_file: crates/squawk_parser/test_data/ok/alter_default_privileges.sql
---
SOURCE_FILE
  COMMENT "-- simple"
  WHITESPACE "\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    SELECT_KW "select"
    WHITESPACE "\n    "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n    "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- full"
  WHITESPACE "\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    FOR_KW "for"
    WHITESPACE " "
    ROLE_KW "role"
    WHITESPACE " "
    IDENT "r"
    COMMA ","
    WHITESPACE " "
    CURRENT_USER_KW "current_user"
    WHITESPACE "\n  "
    IN_KW "in"
    WHITESPACE " "
    SCHEMA_KW "schema"
    WHITESPACE " "
    NAME_REF
      IDENT "s"
    COMMA ","
    WHITESPACE " "
    NAME_REF
      IDENT "s2"
    COMMA ","
    WHITESPACE " "
    NAME_REF
      IDENT "s3"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    SELECT_KW "select"
    WHITESPACE "\n    "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n    "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    FOR_KW "for"
    WHITESPACE " "
    USER_KW "user"
    WHITESPACE " "
    SESSION_USER_KW "session_user"
    WHITESPACE "\n  "
    IN_KW "in"
    WHITESPACE " "
    SCHEMA_KW "schema"
    WHITESPACE " "
    NAME_REF
      IDENT "s"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    SELECT_KW "select"
    WHITESPACE "\n    "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n    "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- grant_tables"
  WHITESPACE "\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    SELECT_KW "select"
    WHITESPACE "\n    "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n    "
    TO_KW "to"
    WHITESPACE " "
    GROUP_KW "group"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    INSERT_KW "insert"
    WHITESPACE "\n    "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n    "
    TO_KW "to"
    WHITESPACE " "
    GROUP_KW "group"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    UPDATE_KW "update"
    WHITESPACE "\n    "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n    "
    TO_KW "to"
    WHITESPACE " "
    IDENT "public"
    COMMA ","
    WHITESPACE " "
    CURRENT_USER_KW "current_user"
    COMMA ","
    WHITESPACE " "
    GROUP_KW "group"
    WHITESPACE " "
    SESSION_USER_KW "session_user"
    WHITESPACE "\n    "
    WITH_KW "with"
    WHITESPACE " "
    GRANT_KW "grant"
    WHITESPACE " "
    OPTION_KW "option"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    DELETE_KW "delete"
    WHITESPACE "\n    "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n    "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    TRUNCATE_KW "truncate"
    WHITESPACE "\n    "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n    "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    REFERENCES_KW "references"
    WHITESPACE "\n    "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n    "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    TRIGGER_KW "trigger"
    WHITESPACE "\n    "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n    "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    IDENT "maintain"
    WHITESPACE "\n    "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n    "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    SELECT_KW "select"
    COMMA ","
    WHITESPACE " "
    INSERT_KW "insert"
    COMMA ","
    WHITESPACE " "
    UPDATE_KW "update"
    COMMA ","
    WHITESPACE " "
    DELETE_KW "delete"
    COMMA ","
    WHITESPACE " "
    TRUNCATE_KW "truncate"
    COMMA ","
    WHITESPACE " "
    REFERENCES_KW "references"
    COMMA ","
    WHITESPACE " "
    TRIGGER_KW "trigger"
    COMMA ","
    WHITESPACE " "
    IDENT "maintain"
    WHITESPACE "\n    "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n    "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    ALL_KW "all"
    WHITESPACE "\n    "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n    "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    ALL_KW "all"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n    "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n    "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- grant_sequences"
  WHITESPACE "\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    IDENT "usage"
    COMMA ","
    WHITESPACE " "
    SELECT_KW "select"
    COMMA ","
    WHITESPACE " "
    UPDATE_KW "update"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    SEQUENCES_KW "sequences"
    WHITESPACE "\n  "
    TO_KW "to"
    WHITESPACE " "
    IDENT "u"
    COMMA ","
    WHITESPACE " "
    GROUP_KW "group"
    WHITESPACE " "
    CURRENT_USER_KW "current_user"
    COMMA ","
    WHITESPACE " "
    GROUP_KW "group"
    WHITESPACE " "
    SESSION_USER_KW "session_user"
    COMMA ","
    WHITESPACE " "
    GROUP_KW "group"
    WHITESPACE " "
    CURRENT_ROLE_KW "current_role"
    WHITESPACE "\n  "
    WITH_KW "with"
    WHITESPACE " "
    GRANT_KW "grant"
    WHITESPACE " "
    OPTION_KW "option"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    IDENT "usage"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    SEQUENCES_KW "sequences"
    WHITESPACE "\n  "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- grant_functions"
  WHITESPACE "\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    EXECUTE_KW "execute"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    FUNCTIONS_KW "functions"
    WHITESPACE "\n  "
    TO_KW "to"
    WHITESPACE " "
    IDENT "u"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    ALL_KW "all"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    ROUTINES_KW "routines"
    WHITESPACE "\n  "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    ALL_KW "all"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    ROUTINES_KW "routines"
    WHITESPACE "\n  "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
    WHITESPACE "\n  "
    WITH_KW "with"
    WHITESPACE " "
    GRANT_KW "grant"
    WHITESPACE " "
    OPTION_KW "option"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- grant_types"
  WHITESPACE "\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    IDENT "usage"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    TYPES_KW "types"
    WHITESPACE "\n  "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    ALL_KW "all"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    TYPES_KW "types"
    WHITESPACE "\n  "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
    COMMA ","
    WHITESPACE " "
    GROUP_KW "group"
    WHITESPACE " "
    IDENT "u"
    COMMA ","
    WHITESPACE " "
    IDENT "public"
    WHITESPACE "\n  "
    WITH_KW "with"
    WHITESPACE " "
    GRANT_KW "grant"
    WHITESPACE " "
    OPTION_KW "option"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- grant_schemas"
  WHITESPACE "\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    IDENT "usage"
    COMMA ","
    WHITESPACE " "
    CREATE_KW "create"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    SCHEMAS_KW "schemas"
    WHITESPACE "\n  "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    GRANT_KW "grant"
    WHITESPACE " "
    ALL_KW "all"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    SCHEMAS_KW "schemas"
    WHITESPACE "\n  "
    TO_KW "to"
    WHITESPACE " "
    IDENT "r"
    COMMA ","
    WHITESPACE " "
    GROUP_KW "group"
    WHITESPACE " "
    IDENT "u"
    COMMA ","
    WHITESPACE " "
    IDENT "public"
    WHITESPACE "\n  "
    WITH_KW "with"
    WHITESPACE " "
    GRANT_KW "grant"
    WHITESPACE " "
    OPTION_KW "option"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- revoke_tables"
  WHITESPACE "\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    REVOKE_KW "revoke"
    WHITESPACE " "
    GRANT_KW "grant"
    WHITESPACE " "
    OPTION_KW "option"
    WHITESPACE " "
    FOR_KW "for"
    WHITESPACE "\n  "
    SELECT_KW "select"
    COMMA ","
    WHITESPACE " "
    INSERT_KW "insert"
    COMMA ","
    WHITESPACE " "
    UPDATE_KW "update"
    COMMA ","
    WHITESPACE " "
    DELETE_KW "delete"
    COMMA ","
    WHITESPACE " "
    TRUNCATE_KW "truncate"
    COMMA ","
    WHITESPACE " "
    REFERENCES_KW "references"
    COMMA ","
    WHITESPACE " "
    TRIGGER_KW "trigger"
    COMMA ","
    WHITESPACE " "
    IDENT "maintain"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n  "
    FROM_KW "from"
    WHITESPACE " "
    IDENT "r"
    COMMA ","
    WHITESPACE " "
    GROUP_KW "group"
    WHITESPACE " "
    CURRENT_USER_KW "current_user"
    WHITESPACE "\n  "
    CASCADE_KW "cascade"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    REVOKE_KW "revoke"
    WHITESPACE " "
    SELECT_KW "select"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    TABLES_KW "tables"
    WHITESPACE "\n  "
    FROM_KW "from"
    WHITESPACE " "
    IDENT "r"
    WHITESPACE "\n  "
    RESTRICT_KW "restrict"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- revoke_sequences"
  WHITESPACE "\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    REVOKE_KW "revoke"
    WHITESPACE " "
    GRANT_KW "grant"
    WHITESPACE " "
    OPTION_KW "option"
    WHITESPACE " "
    FOR_KW "for"
    WHITESPACE "\n  "
    IDENT "usage"
    COMMA ","
    WHITESPACE " "
    SELECT_KW "select"
    COMMA ","
    WHITESPACE " "
    UPDATE_KW "update"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    SEQUENCES_KW "sequences"
    WHITESPACE "\n  "
    FROM_KW "from"
    WHITESPACE " "
    IDENT "r"
    COMMA ","
    WHITESPACE " "
    GROUP_KW "group"
    WHITESPACE " "
    CURRENT_USER_KW "current_user"
    WHITESPACE "\n  "
    CASCADE_KW "cascade"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    REVOKE_KW "revoke"
    WHITESPACE " "
    SELECT_KW "select"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    SEQUENCES_KW "sequences"
    WHITESPACE "\n  "
    FROM_KW "from"
    WHITESPACE " "
    IDENT "r"
    WHITESPACE "\n  "
    RESTRICT_KW "restrict"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- revoke_functions"
  WHITESPACE "\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    REVOKE_KW "revoke"
    WHITESPACE " "
    GRANT_KW "grant"
    WHITESPACE " "
    OPTION_KW "option"
    WHITESPACE " "
    FOR_KW "for"
    WHITESPACE "\n  "
    EXECUTE_KW "execute"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    FUNCTIONS_KW "functions"
    WHITESPACE "\n  "
    FROM_KW "from"
    WHITESPACE " "
    IDENT "r"
    COMMA ","
    WHITESPACE " "
    GROUP_KW "group"
    WHITESPACE " "
    CURRENT_USER_KW "current_user"
    WHITESPACE "\n  "
    CASCADE_KW "cascade"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    REVOKE_KW "revoke"
    WHITESPACE " "
    ALL_KW "all"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    ROUTINES_KW "routines"
    WHITESPACE "\n  "
    FROM_KW "from"
    WHITESPACE " "
    IDENT "r"
    WHITESPACE "\n  "
    RESTRICT_KW "restrict"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- revoke_types"
  WHITESPACE "\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    REVOKE_KW "revoke"
    WHITESPACE " "
    GRANT_KW "grant"
    WHITESPACE " "
    OPTION_KW "option"
    WHITESPACE " "
    FOR_KW "for"
    WHITESPACE "\n  "
    IDENT "usage"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    TYPES_KW "types"
    WHITESPACE "\n  "
    FROM_KW "from"
    WHITESPACE " "
    IDENT "r"
    COMMA ","
    WHITESPACE " "
    GROUP_KW "group"
    WHITESPACE " "
    CURRENT_USER_KW "current_user"
    WHITESPACE "\n  "
    CASCADE_KW "cascade"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    REVOKE_KW "revoke"
    WHITESPACE " "
    ALL_KW "all"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    TYPES_KW "types"
    WHITESPACE "\n  "
    FROM_KW "from"
    WHITESPACE " "
    IDENT "r"
    WHITESPACE "\n  "
    RESTRICT_KW "restrict"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  COMMENT "-- revoke_schemas"
  WHITESPACE "\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    REVOKE_KW "revoke"
    WHITESPACE " "
    GRANT_KW "grant"
    WHITESPACE " "
    OPTION_KW "option"
    WHITESPACE " "
    FOR_KW "for"
    WHITESPACE "\n  "
    IDENT "usage"
    COMMA ","
    WHITESPACE " "
    CREATE_KW "create"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    SCHEMAS_KW "schemas"
    WHITESPACE "\n  "
    FROM_KW "from"
    WHITESPACE " "
    IDENT "r"
    COMMA ","
    WHITESPACE " "
    GROUP_KW "group"
    WHITESPACE " "
    CURRENT_USER_KW "current_user"
    WHITESPACE "\n  "
    CASCADE_KW "cascade"
  SEMICOLON ";"
  WHITESPACE "\n\n"
  ALTER_DEFAULT_PRIVILEGES_STMT
    ALTER_KW "alter"
    WHITESPACE " "
    DEFAULT_KW "default"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    REVOKE_KW "revoke"
    WHITESPACE " "
    ALL_KW "all"
    WHITESPACE " "
    PRIVILEGES_KW "privileges"
    WHITESPACE "\n  "
    ON_KW "on"
    WHITESPACE " "
    SCHEMAS_KW "schemas"
    WHITESPACE "\n  "
    FROM_KW "from"
    WHITESPACE " "
    IDENT "r"
    WHITESPACE "\n  "
    RESTRICT_KW "restrict"
  SEMICOLON ";"
  WHITESPACE "\n\n"
